### PR TITLE
RavenDB-12524 - VoronUnrecoverableErrorException when the disk is full

### PR DESCRIPTION
**Pull request by grisha-kotler**
http://issues.hibernatingrhinos.com/issue/RavenDB-12524

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 1]
```diff
@@ -1573,12 +1598,15 @@ public bool DoIndexingWork(IndexingStatsScope stats, CancellationToken cancellat
...
                             SlowWriteNotification.Notify(commitStats, DocumentDatabase);
-
                             stats.RecordCommitStats(commitStats.NumberOfModifiedPages, commitStats.NumberOf4KbsWrittenToDisk);
                         }
                     }
                     catch
                     {
+                        // need to call it here to the let the index continue running
+                        // we'll stop when we reach the index error threshold 
+                        _mre.Set();
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Index.cs]: This is called on _any_ exception, but isn't the intent just for disk full?

File: src/Raven.Server/Indexing/LuceneVoronDirectory.cs [hunk 2]
```diff
@@ -127,6 +126,12 @@ public override void DeleteFile(string name, IState s)
             if (readResult == null)
                 throw new FileNotFoundException("Could not find file", name);
 
+            if (_indexOutputFilesSummary.HasVoronWriteErrors)
```
Comment (ayende) [inline src/Raven.Server/Indexing/LuceneVoronDirectory.cs]: Move this to the top of the method.
Probably a good idea to do this on the Write as well

Comment (ayende) [reply src/Raven.Server/Indexing/LuceneVoronDirectory.cs]: Also, on Dispose, so it wouldn't try to create a new stream on another file?

File: src/Raven.Server/Indexing/VoronIndexOutput.cs [hunk 3]
```diff
@@ -73,11 +114,54 @@ protected override void Dispose(bool disposing)
...
                 PosixFile.DeleteOnClose(_fileTempPath);
             }
         }
+
+        private void ThrowDiskFullException()
+        {
+            var directory = Path.GetDirectoryName(_fileTempPath);
+            var driveInfo = DiskSpaceChecker.GetDiskSpaceInfo(directory);
+            throw new DiskFullException($"There isn't enough space to flush the buffer of the file: {_fileTempPath}. " +
+                                        $"Currently available space: {driveInfo?.TotalFreeSpace}");
```
Comment (ayende) [inline src/Raven.Server/Indexing/VoronIndexOutput.cs]: Handle the case where this is null and say `space: N/A` or something

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Indexing/LuceneVoronDirectory.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Indexing/LuceneVoronDirectory.cs]: done

File: src/Raven.Server/Indexing/VoronIndexOutput.cs (hunk 3 above)
Comment (grisha-kotler) [reply src/Raven.Server/Indexing/VoronIndexOutput.cs]: done

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/Index.cs]: This is for any exception. We have error handling and after a couple of errors we set the index as errored. Unless we do that the index just stops

Comment (arekpalinski) [reply src/Raven.Server/Documents/Indexes/Index.cs]: This is okay for me but I'm a bit worried that we'll error indexes very quickly after getting any error here  (e.g. what is the chance we'll recover from the disk full exception, we'll repeat it 10 times and mark the index as errored).

I wonder if we should have a notion of fallback mode - e.g. after getting end of disk space then wait 10 seconds, and run indexing then.. next wait 20 seconds etc until the max fallback mode duration.

Comment (ayende) [reply src/Raven.Server/Documents/Indexes/Index.cs]: Sounds good to me.
We should also force a cleanup of our state in this case, so a flush / sync of Voron and then ask to remove all the old recylced journals, old scratches, etc.

**Comment by grisha-kotler** [comment]
Done

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/Index.cs]: done

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 4]
```diff
@@ -1385,23 +1389,51 @@ internal void HandleExcessiveNumberOfReduceErrors(IndexingStatsScope stats, Exce
...
-            if (State == IndexState.Error || diskFullErrors < DiskFullErrorLimit)
+                // force flush and sync
+                GlobalFlushingBehavior.GlobalFlusher.Value.MaybeFlushEnvironment(storageEnvironment);
+                if (_logsAppliedEvent.Wait(timeToWaitInSeconds))
+                {
+                    storageEnvironment.ForceSyncDataFile();
+                }
+
+                // wait for sync
+                Task.Delay(1000 * timeToWaitInSeconds, _indexingProcessCancellationTokenSource.Token).Wait();
```
Comment (ayende) [inline src/Raven.Server/Documents/Indexes/Index.cs]: Why are you doing this? We are effectively waiting twice.
In any case, worse case we can try again and we'll wait for the sync on the next iteration, no?

Comment (grisha-kotler) [reply src/Raven.Server/Documents/Indexes/Index.cs]: fixed the calculations

**Comment by arekpalinski** [comment]
@grisha-kotler Please note failing test: http://jenkins.hibernatingrhinos.com/job/PR_Tests_v4.1/2280/

**Comment by arekpalinski** [comment]
We have it reported as failing occasionally: http://issues.hibernatingrhinos.com/issue/RavenDB-12325

**Pull request merged**